### PR TITLE
[Project] coroutine threaded model for API calls: thread pool

### PR DESCRIPTION
**Pull request by negram**

### CODE REVIEW

File: src/lua/lua_config.c [hunk 1]
```diff
@@ -1270,35 +1299,40 @@ lua_metric_symbol_callback (struct rspamd_task *task, gpointer ud)
...
 					}
 				}
 
 			}
 
-			lua_pop (L, nresults);
+			lua_pop (thread, nresults);
 		}
+
+		assert (lua_gettop (thread) == cd->stack_level); /* we properly cleaned up the stack */
```
Comment (vstakhov) [inline src/lua/lua_config.c]: Please change to `g_assert`.

File: src/lua/lua_config.c [hunk 2]
```diff
@@ -1184,43 +1187,69 @@ lua_watcher_callback (gpointer session_data, gpointer ud)
...
+
+		tb = lua_touserdata (thread, -1);
 		msg_err_task ("call to (%s) failed (%d): %v", cd->symbol, ret, tb);
 
 		if (tb) {
 			g_string_free (tb, TRUE);
-			lua_pop (L, 1);
+			lua_pop (thread, 1);
 		}
+		assert (lua_gettop (thread) >= cd->stack_level);
```
Comment (vstakhov) [inline src/lua/lua_config.c]: `g_assert`

File: src/lua/lua_config.c [hunk 3]
```diff
@@ -1184,43 +1187,69 @@ lua_watcher_callback (gpointer session_data, gpointer ud)
...
+	ptask = lua_newuserdata (thread, sizeof (struct rspamd_task *));
+	rspamd_lua_setclass (thread, "rspamd{task}", -1);
 	*ptask = task;
 
-	if ((ret = lua_pcall (L, 1, LUA_MULTRET, err_idx)) != 0) {
-		tb = lua_touserdata (L, -1);
+	ret = lua_resume (thread, 1);
+
+	if (ret == LUA_YIELD) {
+		msg_err_task ("LUA_YIELD");
```
Comment (vstakhov) [inline src/lua/lua_config.c]: Why err here?

File: src/lua/lua_config.c [hunk 4]
```diff
@@ -1184,43 +1187,69 @@ lua_watcher_callback (gpointer session_data, gpointer ud)
...
 	struct rspamd_task **ptask;
-	gint level = lua_gettop (cd->L), nresults, err_idx, ret;
-	lua_State *L = cd->L;
-	GString *tb;
-	struct rspamd_symbol_result *s;
+	gint ret;
 
-	lua_pushcfunction (L, &rspamd_lua_traceback);
-	err_idx = lua_gettop (L);
+	struct thread_entry *thread_entry = lua_thread_pool_get(task->cfg->lua_thread_pool);
```
Comment (vstakhov) [inline src/lua/lua_config.c]: Space before `(` is missing.

File: src/lua/lua_config.c [hunk 5]
```diff
@@ -1184,43 +1187,69 @@ lua_watcher_callback (gpointer session_data, gpointer ud)
...
 
-	ptask = lua_newuserdata (L, sizeof (struct rspamd_task *));
-	rspamd_lua_setclass (L, "rspamd{task}", -1);
+	ptask = lua_newuserdata (thread, sizeof (struct rspamd_task *));
+	rspamd_lua_setclass (thread, "rspamd{task}", -1);
 	*ptask = task;
 
-	if ((ret = lua_pcall (L, 1, LUA_MULTRET, err_idx)) != 0) {
-		tb = lua_touserdata (L, -1);
+	ret = lua_resume (thread, 1);
```
Comment (vstakhov) [inline src/lua/lua_config.c]: I was not sure about `nresults=1` here. Shouldn't it be `LUA_MULTRET`?

File: src/lua/lua_config.c [hunk 6]
```diff
@@ -1184,43 +1187,69 @@ lua_watcher_callback (gpointer session_data, gpointer ud)
...
+	GString *tb;
+	struct lua_callback_data *cd = ud;
+	int nresults;
+	struct rspamd_symbol_result *s;
+	struct thread_entry *thread_entry = cd->thread_entry;
+	lua_State *thread = thread_entry->lua_state;
+
+	if (ret != 0) {
+		lua_pushcfunction (thread, rspamd_lua_traceback);
+		lua_call (thread, 0, 1);
```
Comment (vstakhov) [inline src/lua/lua_config.c]: Same here: why nresults is equal to 1?

File: src/lua/lua_thread_pool.c [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+	g_free (ent);
+}
+
+struct lua_thread_pool *
+lua_thread_pool_new (lua_State * L)
+{
+	struct lua_thread_pool * pool = g_new0 (struct lua_thread_pool, 1);
+
+	pool->L = L;
+	pool->max_items = 100;
```
Comment (vstakhov) [inline src/lua/lua_thread_pool.c]: We might probably want to adjust this.

File: src/lua/lua_thread_pool.c [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+	pool->available_items = g_queue_new ();
+	int i;
+
+	struct thread_entry *ent;
+	for (i = 0; i < pool->max_items; i ++) {
```
Comment (vstakhov) [inline src/lua/lua_thread_pool.c]: Why do you preallocate all threads in creation? Wouldn't it be more wise to preallocate some small amount of threads (e.g. 8) and then allocate them in a lazy matter?

File: src/lua/lua_thread_pool.c [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+	gpointer cur;
+	struct thread_entry *ent = NULL;
+
+	cur = g_queue_pop_head (pool->available_items);
+
+	if (cur) {
+		ent = cur;
+	}
+	else {
+		ent = thread_entry_new (pool->L);
```
Comment (vstakhov) [inline src/lua/lua_thread_pool.c]: In this case thread is created but not pushed to the queue. It looks strange.

File: src/lua/lua_thread_pool.c [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+	}
+	return ent;
+}
+
+void
+lua_thread_pool_return(struct lua_thread_pool *pool, struct thread_entry *thread_entry)
+{
+	assert (lua_status (thread_entry->lua_state) == 0); // we can't return a running/yielded stack into the pool
```
Comment (vstakhov) [inline src/lua/lua_thread_pool.c]: `g_assert`

File: src/lua/lua_thread_pool.h [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+#include <lua.h>
+
+struct thread_entry {
+	lua_State *lua_state;
+	gint thread_index;
+};
+
+struct thread_pool;
+
+struct lua_thread_pool *
```
Comment (vstakhov) [inline src/lua/lua_thread_pool.h]: Could you please add doxygen like comments for this public API?

File: src/lua/lua_thread_pool.c [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+	struct thread_entry *running_entry;
+};
+
+static struct thread_entry *
+thread_entry_new (lua_State * L)
+{
+	struct thread_entry *ent;
+	ent = g_malloc (sizeof *ent);
+	ent->lua_state = lua_newthread (L);
+	ent->thread_index = luaL_ref (L, LUA_REGISTRYINDEX);
```
Comment (vstakhov) [inline src/lua/lua_thread_pool.c]: Please add an empty line before return statement.

File: src/lua/lua_thread_pool.c [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+	struct thread_entry *ent = NULL;
+
+	cur = g_queue_pop_head (pool->available_items);
+
+	if (cur) {
+		ent = cur;
+	}
+	else {
+		ent = thread_entry_new (pool->L);
+	}
```
Comment (vstakhov) [inline src/lua/lua_thread_pool.c]: Empty line missing.

File: src/lua/lua_thread_pool.c [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+	return ent;
+}
+
+void
+lua_thread_pool_return(struct lua_thread_pool *pool, struct thread_entry *thread_entry)
+{
+	assert (lua_status (thread_entry->lua_state) == 0); // we can't return a running/yielded stack into the pool
+	if (pool->running_entry == thread_entry) {
+		pool->running_entry = NULL;
+	}
```
Comment (vstakhov) [inline src/lua/lua_thread_pool.c]: I'd suggest to add an empty line here.

File: src/lua/lua_thread_pool.c [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+void
+lua_thread_pool_terminate_entry(struct lua_thread_pool *pool, struct thread_entry *thread_entry)
+{
+	struct thread_entry *ent = NULL;
+
+	if (pool->running_entry == thread_entry) {
+		pool->running_entry = NULL;
+	}
+
+	// we should only terminate failed threads
```
Comment (vstakhov) [inline src/lua/lua_thread_pool.c]: Please use C style comments `/* */` to keep the general style consistent.

File: src/lua/lua_thread_pool.c (hunk 9 above)
Comment (negram) [reply src/lua/lua_thread_pool.c]: The thread will be pushed into the queue once it's freed. The queue contains only available threads, while this functions makes thread unavailable.

**Pull request merged**